### PR TITLE
Refactor: transform submodule; Field storage control; CondSRF optimization

### LINKED ISSUE #195: Rework field transformations workflow
### Discussed in https://github.com/GeoStat-Framework/GSTools/discussions/168

At the moment, transformations work on a given field and modify the `field` attribute inplace.

Ideas:
- return the transformed field
- add a selector to specify the field attribute name (`"field"` by default)
- add `store` keyword (`True` by default to modify inplace, `False` to only return, `<str>` to store in other attribute given by name)

### Example
```python
import gstools as gs
x = y = range(100)
model = gs.Gaussian(dim=2, var=1, len_scale=10)
srf = gs.SRF(model)
srf.structured([x, y])
log_field = gs.transform.normal_to_lognormal(srf, field="field", store="log_field")
srf.plot(field="log_field")
print(srf.field)
print(srf.log_field)
```
This has the advantage, that the original field is still available and the new field is also stored in the class.
The plotting routines are already able to select a field by name.

We should use the `Field.post_field` method to store the fields by name.

### PR DESCRIPTION
**Pull request by MuellerSeb**
Closes #195 #196 

- `gstools.transform`:
  - add keywords `field`, `store` and `process` keyword to all transformations to control storage and respect `normalizer`
  - added `apply_function` transformation
  - added `apply` as wrapper for all transformations
  - added `transform` method to all `Field` (sub)classes as interface to `transform.apply`
  - added checks for normal fields to work smoothly with recently added `normalizer` submodule
- `Field`:
  - allow naming fields when generating and control storage with `store` keyword
  - all subclasses now have the `post_process` keyword (apply mean, normalizer, trend)
  - added subscription to access fields by name (`Field["field"]`)
  - added `set_pos` method to set position tuple
  - allow reusing of present `pos` tuple
  - added `pos`, `mesh_type`, `field_names`, `field_shape`, `all_fields` properties
- `CondSRF`:
  - memory optimization by forwarding `pos` to underlying `krige` instance
  - only recalculate kriging field if the `pos` tuple changed (optimized ensemble generation)
- performance improvement by using `np.asarray` instead of `np.array` where possible
- updated examples to use new features

### CODE REVIEW

**Comment by MuellerSeb** [comment]
@LSchueler this is ready for review.

File: gstools/field/base.py [hunk 1]
```diff
@@ -343,6 +470,134 @@ def plot(
...
+        mesh_type : :class:`str`, optional
+            'structured' / 'unstructured'
+            Default: `"unstructured"`
+        info : :class:`bool`, optional
+            Whether to return information
+
+        Returns
+        -------
+        info : :class:`dict`
+            Information about settings.
```
Comment (LSchueler) [inline gstools/field/base.py:491]: I think a warning or note field in the docs is appropriate here, to warn users that the current fields will get overwritten.

File: gstools/field/cond_srf.py [hunk 2]
```diff
@@ -143,6 +192,54 @@ def set_generator(self, generator, **generator_kwargs):
...
+            'structured' / 'unstructured'
+            Default: `"unstructured"`
+        info : :class:`bool`, optional
+            Whether to return information
+
+        Returns
+        -------
+        info : :class:`dict`
+            Information about settings.
+        """
```
Comment (LSchueler) [inline gstools/field/cond_srf.py:214]: See warnings or note comment.

File: gstools/transform/array.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+
+def array_boxcox(field, lmbda=1, shift=0):
+    """
+    (Inverse) Box-Cox transformation to denormalize data.
+
+    After this transformation, the again Box-Cox transformed field is normal
+    distributed.
+
+    See: https://en.wikipedia.org/wiki/Power_transform#Box%E2%80%93Cox_transformation
```
Comment (LSchueler) [inline gstools/transform/array.py:116]: Would it be silly to add a date when the link was last checked?

**Review by LSchueler — CHANGES_REQUESTED** [review]
Wow! Awesome!
Just a few minor comments in the code, as usual.

And I fixed a typo and some forgotten "optional" descriptors for the now optional `pos` argument. So, don't forget to pull.

File: examples/01_random_field/01_srf_ensemble.py [hunk 4]
```diff
@@ -14,26 +16,26 @@
...
-# members, for better visualisation and save them in a list and in a first
+# members, for better visualisation, save them in to srf class and in a first
 # step, we will be using the loop counter as the seeds.
 
-
 ens_no = 4
-field = []
 for i in range(ens_no):
-    field.append(srf.structured([x, y], seed=i))
+    srf(seed=i, store=f"field{i}")
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply examples/01_random_field/01_srf_ensemble.py:29]: I would keep this as an educational example. As you said: We introduce better seeding afterwards.

File: gstools/field/base.py [hunk 5]
```diff
@@ -70,10 +90,10 @@ def __init__(
         dim=None,
     ):
         # initialize attributes
-        self.pos = None
-        self.mesh_type = None
-        self.field = None
-        # initialize private attributes
+        self._mesh_type = "unstructured"  # default
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply gstools/field/base.py:93]: but true!

File: gstools/field/base.py [hunk 6]
```diff
@@ -104,24 +168,33 @@ def __call__(
         post_process : :class:`bool`, optional
             Whether to apply mean, normalizer and trend to the field.
             Default: `True`
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply gstools/field/base.py:170]: I would keep the "whether" formulations. I "learned" this during a contribution for another package, where the maintainer demanded to have this formulation with all switches. You can come up with a PR if you'd like to change this.

File: gstools/transform/array.py (hunk 3 above)
Comment (MuellerSeb) [reply gstools/transform/array.py:116]: I don't really want to pollute the docs with such information. I don't think this wiki page will change much in the future ;-)

File: gstools/field/cond_srf.py (hunk 2 above)
Comment (MuellerSeb) [reply gstools/field/cond_srf.py]: Done.

File: gstools/field/base.py (hunk 1 above)
Comment (MuellerSeb) [reply gstools/field/base.py:496]: Done.

File: examples/06_conditioned_fields/00_condition_ensemble.py [hunk 7]
```diff
@@ -26,15 +26,18 @@
...
 cond_srf = gs.CondSRF(krige)
+cond_srf.set_pos(gridx)
 
 ###############################################################################
+# We can specify individual names for each field by the keyword `store`:
 
-fields = []
 for i in range(100):
-    fields.append(cond_srf(gridx, seed=i))
+    cond_srf(seed=i, store=f"f{i}")
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply examples/06_conditioned_fields/00_condition_ensemble.py]: Done.

File: examples/06_conditioned_fields/01_2D_condition_ensemble.py [hunk 8]
```diff
@@ -20,34 +20,39 @@
...
+# By specifying ``store=[f"fld{i}", False, False]``, only the conditioned field
+# is stored with the specified name. The raw random field and the raw kriging
+# field is not stored. This way, we can access each conditioned field by index
+# ``cond_srf[i]``:
 
 ens_no = 4
-field = []
 for i in range(ens_no):
-    field.append(cond_srf.structured([x, y], seed=i))
+    cond_srf(seed=i, store=[f"fld{i}", False, False])
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply examples/06_conditioned_fields/01_2D_condition_ensemble.py]: Done.

**Pull request merged**